src/SourceFile.js

```diff
@@ -0,0 +1,229 @@
+/*
+ * SourceFile.js - Represent an ilib-lint rule
+ *
+ * Copyright © 2024 JEDLSoft
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import fs from 'fs';
+import NotImplementedError from "./NotImplementedError.js";
+
+/**
+ * @class Represent a source file that can be linted in some way.
+ */
+class SourceFile {
+    /**
+     * Construct a new source file instance. Parsers should produce these.
+     *
+     * @param {String} uri URI or path to the source file
+     * @param {Object} [options] options to the constructor
+     * @param {String} [options.sourceLocale] the source locale of the files
+     * being linted
+     * @param {Function} [options.getLogger] a callback function provided by
+     * the linter to retrieve the log4js logger
+     * @param [String] [options.type] the type of this file
+     * @constructor
+     */
+    constructor(uri, options) {
+        if (!uri) {
+            throw new Error("Attempt to create a SourceFile without a file path");
+        }
+        this.filePath = uri;
+        this.logger = (typeof(options.getLogger) === "function") ? options.getLogger("ilib-lint-common.SourceFile") : undefined;
+        this.sourceLocale = options?.sourceLocale || "en-US";
+        this.type = options?.type || "";
+    }
+
+    /**
+     * Read the contents of the file into memory.
+     * @throws Error if the file could not be read, or if it is not encoded in UTF-8
+     * @protected
+     */
+    read() {
+        // do not convert to a string for the raw property
+        this.raw = fs.readFileSync(this.filePath);
+
+        // ... but do convert for the string property
+        this.content = this.raw.toString("utf-8");
+
```

Comment:
Having two copies of the file content in memory seems weird - is this on purpose?